src/index.js

```diff
@@ -1,174 +1,231 @@
 'use strict'
 
-import createWildcard from './wild'
+import createWildcard from './wildcard'
 
 const isArray = Array.isArray
 const keys = Object.keys
 
-export default options => {
+export default (options = {}) => {
   const wild = createWildcard({ enabled: options.wildcard })
 
-  let rooms = {}
-  let sids = {}
+  let rooms = new Map();
+  let sids = new Map();
 
   /**
-   * Get rooms socket is subscribed to.
-   *
-   * @param {String} id
-   * @param {Function} [fn]
+   * Get rooms socket is subscribed to. Or all rooms in the server
+   * Used for get rooms
+   * @param {String} [id] - spark's ID
+   * @returns {Promise} containing rooms the s
    * @api public
    */
-
-  function get(id, fn = noop) {
-    let rms = keys(id ? sids[id] || {} : rooms)
-    setImmediate(fn, null, rms)
+  function get(id) {
+    if (id) {
+      return Promise.resolve(Array.from(sids.get(id)));
+    }
+    let roomsArray = [];
+    rooms.forEach((value, key) => {
+      roomsArray.push(key);
+    });
+    return Promise.resolve(roomsArray);
   }
 
   /**
    * Add a socket to a room.
-   *
-   * @param {String} id
-   * @param {String} room
-   * @param {Function} [fn]
+   * Used for joining rooms
+   * @param {String} id - spark's ID
+   * @param {String} room - the room to join
    * @api public
    */
 
-  function add(id, room, fn = noop) {
-    sids[id] = sids[id] || []
-    sids[id][room] = true
-    rooms[room] = rooms[room] || {}
-    rooms[room][id] = true
-    wild.add(room)
-    setImmediate(fn, null, null)
+  function add(id, room) {
+    let socketRooms = sids.get(id);
+    let socketsInRoom = rooms.get(room);
+    // if socket has a rooms set just add the room
+    if (socketRooms instanceof Set) {
+      socketRooms.add(room);
+    } else {
+      // socket has no rooms set, create and add the room
+      sids.set(id, new Set([room]))
+    }
+    // if room has a set of sockets, just add the new socket id
+    if (socketsInRoom instanceof Set) {
+      socketsInRoom.add(id);
+    } else {
+      // room has no set of sockets, create and add it
+      rooms.set(room, new Set([id]));
+    }
+    // add to wild rooms
+    // TODO visit wild logic to work Map and Set
+    // wild.add(room);
+    return Promise.resolve({success: true});
```

Comment:
Hmm not really pleased with the resolve value but I'm also not sure what to return. How about nothing (`undefined`)?